dask/base.py

```diff
@@ -211,7 +211,24 @@ def is_dask_collection(x) -> bool:
 
     """
     try:
-        return x.__dask_graph__() is not None
+        if "xarray" in type(x).__module__:
+            import xarray
+
+            if isinstance(x, xarray.Dataset):
+                return any(is_dask_collection(v._data) for _, v in x.variables.items())
+            elif isinstance(x, xarray.DataArray):
+                return is_dask_collection(x.variable._data)
+            elif isinstance(x, xarray.Variable):
+                return is_dask_collection(x._data)
+            else:
+                raise TypeError("Unfamiliar with xarray type", type(x))
+        else:
+            return (
+                hasattr(x, "__dask_graph__")
+                and callable(x.__dask_graph__)
+                and not isinstance(x, type)
+            )
+
```

Comment:
This PR breaks pint, because just like xarray it violates the dask collection protocol by defining a `__dask_graph__()` method that may return None:

```python
>>> import dask.array as da, pint
>>> q = pint.Quantity(1, "m")
>>> q.__dask_graph__()
(None)
>>> q = pint.Quantity(da.zeros(5), "m")
>>> q.__dask_graph__()
HighLevelGraph with 1 layers.
<dask.highlevelgraph.HighLevelGraph object at 0x7f043c3889a0>
 0. zeros_like-902bf3bcabaabb3fedce3f347fb26081
```

There may be more third-party libraries in the wild that do the same.
A blacklist approach would be more robust:

```python
if (
    isinstance(x, type)
    or not hasattr(x, "__dask_graph__")
    or not callable(x.__dask_graph__)
):
    return False

modname = getattr(type(x), "__module__", "")
# Add here other third-party packages where you don't want
# to call the `__dask_graph__` method - typically because it's expensive
if "dask_expr" in modname:
    return True

# xarray, pint, and possibly other libraries return None when they wrap a non-dask object
return x.__dask_graph__() is not None
```

Reply:
The above, however, will mean that we call the __dask_graph__() method of dask_expr.array objects when they're wrapped by xarray or pint. So I would suggest either:
1. hybrid allowlist+blocklist, with final fallback on calling `__dask_graph__()` 
2. change the __dask_graph__() method in dask_expr to be trivial (e.g. write a MutableMapping subclass that is trivial to initialize and is lazy until you actually try to access its contents)